# ==========================================
# PERTEMUAN 2: Struktur Data Python, NumPy & Pandas
# ==========================================

In [189]:
import pandas as pd
import numpy as np

# Struktur Data Python

## List

In [190]:
data = [10, 20, 'Python', True, 3.14]

# Akses & slicing
print("List data", data)
print("Data index ke-0:", data[0]) # → 10
print("Data dengan Index paling belakang:", data[-1]) # → 3.14 (dari belakang)
print("List data ke-1 dan ke-3", data[1:3]) # → (20, 'Python')

# Manipulasi
data.append(99) # tambah di akhir
data.insert(1, 55) # sisipkan di posisi 1
data.remove(20) # hapus nilai pertama = 20
print("\nList yang telah dimanipulasi", data)

List data [10, 20, 'Python', True, 3.14]
Data index ke-0: 10
Data dengan Index paling belakang: 3.14
List data ke-1 dan ke-3 [20, 'Python']

List yang telah dimanipulasi [10, 55, 'Python', True, 3.14, 99]


# Tuple

In [191]:
fruits = ('apple', 'banana', 'cherry')
print(fruits)

('apple', 'banana', 'cherry')


# Dictionary

In [192]:
mahasiswa = {
'nama': 'Budi Santoso',
'nim': '2024001',
'prodi': 'Informatika',
'ipk': 3.85
}

# Akses nilai
print("Dictionary mahasiswa\n", mahasiswa)
print("Nama mahasiswa:", mahasiswa['nama']) # → 'Budi Santoso'
print("Usia mahasiswa:", mahasiswa.get('usia', 0)) # default jika key tak ada

# Update dan hapus
mahasiswa['angkatan'] = 2024 # tambah key baru
del mahasiswa['nim'] # hapus key
print("\nDictionary mahasiswa setelah update dan hapus: \n", mahasiswa)

Dictionary mahasiswa
 {'nama': 'Budi Santoso', 'nim': '2024001', 'prodi': 'Informatika', 'ipk': 3.85}
Nama mahasiswa: Budi Santoso
Usia mahasiswa: 0

Dictionary mahasiswa setelah update dan hapus: 
 {'nama': 'Budi Santoso', 'prodi': 'Informatika', 'ipk': 3.85, 'angkatan': 2024}


## Set

In [193]:
fruits_set = {"apple", "banana", "cherry"}
print(fruits_set)

{'cherry', 'banana', 'apple'}


# Numpy: Komputasi Numerik

## NumPy Array

In [194]:
# Dari list Python
a1D = np.array([1, 2, 3, 4, 5]) # 1D
a2D = np.array([[1,2,3],[4,5,6]]) # 2D

# Array dengan nilai inisial
zeros = np.zeros((3, 4)) # nol semua
ones = np.ones((2, 3)) # satu semua
eye = np.eye(4) # matriks identitas
full = np.full((2, 2), 7) # nilai konstan
print("Zeros\n", zeros, "\nOnes\n", ones, "\nEye\n", eye, "\nFull\n", full)

# Sequence
arange = np.arange(0, 20, 2) # 0,2,4,...18
linsp = np.linspace(0, 1, 11) # 11 titik merata
print("\nArange\n", arange, "\nLinsp\n", linsp)

# Array acak (reproducible)
rng = np.random.default_rng(seed=42)
rand_int = rng.integers(0, 100, (3, 4))
rand_flt = rng.normal(loc=0, scale=1, size=(100,))
print("\nRand Int\n", rand_int, "\nRand Flt\n", rand_flt)

Zeros
 [[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]] 
Ones
 [[1. 1. 1.]
 [1. 1. 1.]] 
Eye
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]] 
Full
 [[7 7]
 [7 7]]

Arange
 [ 0  2  4  6  8 10 12 14 16 18] 
Linsp
 [0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]

Rand Int
 [[ 8 77 65 43]
 [43 85  8 69]
 [20  9 52 97]] 
Rand Flt
 [ 0.1278404  -0.31624259 -0.01680116 -0.85304393  0.87939797  0.77779194
  0.0660307   1.12724121  0.46750934 -0.85929246  0.36875078 -0.9588826
  0.8784503  -0.04992591 -0.18486236 -0.68092954  1.22254134 -0.15452948
 -0.42832782 -0.35213355  0.53230919  0.36544406  0.41273261  0.430821
  2.1416476  -0.40641502 -0.51224273 -0.81377273  0.61597942  1.12897229
 -0.11394746 -0.84015648 -0.82448122  0.65059279  0.74325417  0.54315427
 -0.66550971  0.23216132  0.11668581  0.2186886   0.87142878  0.22359555
  0.67891356  0.06757907  0.2891194   0.63128823 -1.45715582 -0.31967122
 -0.47037265 -0.63887785 -0.27514225  1.49494131 -0.86583112  0.96827835
 -1.68286

## Operasi Vektorisasi dan Fungsi Statistik

In [195]:
# Operasi aritmetika vektorisasi
a = np.array([1, 2, 3, 4, 5])
b = np.array([10, 20, 30, 40, 50])
print(a + b) # → [11 22 33 44 55]
print(a * b) # → [10 40 90 160 250]
print(a ** 2) # → [1 4 9 16 25]

# Universal functions (ufuncs)
np.sqrt(a) # akar kuadrat tiap elemen
np.exp(a) # eksponensial (e^x)
np.log(b) # logaritma natural

# Fungsi statistik
data = rng.normal(0, 1, 1000)
print(f'\nMean: {np.mean(data):.4f}')
print(f'Std: {np.std(data):.4f}')
print(f'Min: {np.min(data):.4f}')
print(f'Max: {np.max(data):.4f}')
print(f'Q1: {np.percentile(data,25):.4f}')
print(f'Q3: {np.percentile(data,75):.4f}')

[11 22 33 44 55]
[ 10  40  90 160 250]
[ 1  4  9 16 25]

Mean: -0.0163
Std: 1.0008
Min: -3.6484
Max: 3.1789
Q1: -0.6897
Q3: 0.6163


# Pandas: Manipulasi DataFrame

## Pandas Series

In [196]:
# Series dari list — index default 0,1,2...
s1 = pd.Series([85, 92, 78, 95, 88])

# Series dengan index kustom
nilai = pd.Series(
[85, 92, 78, 95],
index=['Alice','Bob','Caca','Dina']
)
print("Nilai Bob:", nilai['Bob']) # → 92
print("Filter nilai:\n", nilai[nilai>80]) # filter
print("Nilai rata-rata:", nilai.mean()) # → 87.5
print("\nStatistik:\n", nilai.describe()) # statistik lengkap

Nilai Bob: 92
Filter nilai:
 Alice    85
Bob      92
Dina     95
dtype: int64
Nilai rata-rata: 87.5

Statistik:
 count     4.000000
mean     87.500000
std       7.593857
min      78.000000
25%      83.250000
50%      88.500000
75%      92.750000
max      95.000000
dtype: float64


## Pandas DataFrame

In [197]:
# Membuat DataFrame dari dictionary
data = {
'nama': ['Ali','Budi','Caca','Dina'],
'usia': [22, 25, 21, 23],
'kota': ['Jakarta','Bandung','Surabaya','Medan'],
'ipk': [3.8, 3.5, 3.9, 3.7]
}
df = pd.DataFrame(data)

# Informasi dasar
print("Dimensi data:", df.shape) # → (4, 4)
print("Tipe data per kolom:\n", df.dtypes) # tipe data per kolom
print("Nama Kolom:", df.columns) # nama-nama kolom

Dimensi data: (4, 4)
Tipe data per kolom:
 nama     object
usia      int64
kota     object
ipk     float64
dtype: object
Nama Kolom: Index(['nama', 'usia', 'kota', 'ipk'], dtype='object')


# Praktikum: Dataset Titanic

In [198]:
URL = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

df = pd.read_csv(URL)
df.head()


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## Q1: Baris dan Kolom

In [199]:

df.info()
print("\nJumlah Baris dan Kolom:\n", df.shape)
print("Kolom:\n", df.columns.tolist())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB

Jumlah Baris dan Kolom:
 (891, 12)
Kolom:
 ['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']


Pada dataset ini, berisi **891 baris** dengan **total kolom yang berjumlah 12**. Adapun kolom yang tersedia, yaitu **PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, dan Embarked**.

## Q2: Missing Value

In [200]:
# missing values
missing_count = df.isnull().sum()
print(missing_count[missing_count > 0])

# persentase
missing_pct = (df.isnull().sum() / len(df)) * 100
print("\n", missing_pct[missing_pct > 0])

Age         177
Cabin       687
Embarked      2
dtype: int64

 Age         19.865320
Cabin       77.104377
Embarked     0.224467
dtype: float64


Pada kolom dataset tersebut, ada beberapa yang mengalami missing value. Kolom **Age** mengalami missing value sebanyak **177 baris (19.7%),** kolom **Cabin** sebanyak **687 baris (77.1%),** dan terakhir **Embarked** sebnyak **2 baris (0.2%).**

## Q3: Penumpang Survived

In [201]:
persen_selamat = df['Survived'].mean() * 100
print(f"Persentase yang selamat: {persen_selamat:.2f}%")

Persentase yang selamat: 38.38%


Persentase penumpang yang berhasil selamat **(Survived)** dari dataset ini sebesar **38.38%**.

## Q4: Penumpang Wanita

In [202]:
wk1 = df[(df["Sex"] == "female") & (df["Pclass"] == 1)]
print("Jumlah wanita dari kelas 1:", len(wk1),"\nRata-rata umur wanita dari kelas 1", wk1['Age'].mean())

Jumlah wanita dari kelas 1: 94 
Rata-rata umur wanita dari kelas 1 34.61176470588235


Berdasarkan data, penumpang wanita kelas 1 berjumlah **94 orang** dengan rata-rata berumur **35 tahun**.

## Q5: Tingkat Keselamatan per Kelas

In [203]:
hasil = df.groupby("Pclass").agg(
    total=('PassengerId', 'count'),
    selamat=('Survived', 'mean'),
).round(3)
print(hasil)

        total  selamat
Pclass                
1         216    0.630
2         184    0.473
3         491    0.242


Berdasarkan hasil groupby, **kelas 1** menempati tingkat Survived yang paling banyak dengan rata-rata **63% (216 orang)**. Kemudian disusul oleh **kelas 2** dengan rata-rata **47% (184 orang)** dan terakhir **kelas 3** dengan rata-rata **24% (491 orang)**.

## Q6: Analisis Tambahan

In [204]:
# Membuat kolom FamilySize dan status IsAlone
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1
df['IsAlone'] = df['FamilySize'].apply(lambda x: 1 if x == 1 else 0)

# Analisis tingkat keselamatan
analisis_keluarga = df.groupby('IsAlone').agg(
    total=('PassengerId', 'count'),
    selamat=('Survived', 'mean')
).round(3)

print(analisis_keluarga)

         total  selamat
IsAlone                
0          354    0.506
1          537    0.304


Berdasarkan hasil analisis, penumpang yang bepergian bersama keluarga **(IsAlone=0)** memiliki tingkat keselamatan sebesar **50.6% (354 orang)**. Sedangkan penumpang yang sendirian **(IsAlone=1)** tingkat keselamatannya jauh lebih rendah, yaitu hanya **30.4% (537 orang)**.

Data ini menunjukkan kalau penumpang yang membawa keluarga punya peluang selamat lebih tinggi. Hal ini kemungkinan karena prioritas evakuasi "wanita dan anak-anak" yang biasanya bepergian dalam grup keluarga, atau adanya aksi saling bantu antar anggota keluarga saat mencari sekoci.